### PR TITLE
CUDA: Support for `@overload`

### PR DESCRIPTION
**Pull request by gmarkall**
Joint work with @stuartarchibald.

The implementation consists of:

- Addition of `get_call_template`, `nopython_signatures`, and   `get_overload` to the class that is now `DeviceDispatcher`. These are needed by the overload infrastructure, and are inspired by the equivalents in `_DispatcherBase`.
- Registration of a CUDA `jit` decorator. The registration with `dispatcher_register` is removed and replaced with this new registration, because it did nothing (it was for the `target` kwarg to the `numba.jit` decorator, which is deprecated). The `jit` decorator for overloads is a bit special in that it always adds the `device=True` kwarg to the regular `jit` decorator, because overloads will always be device functions.
- Addition of a `no_cpython_wrapper` kwarg to `jitdevice`, which is needed because the overload infrastructure passes it. No action is required regardless of its value, because there is no CPython wrapper in the CUDA target.

A refactoring is also made: `DeviceFunctionTemplate` is renamed to `DeviceDispatcher`, as it reflects more closely what it is - it is
analagous to the `Dispatcher` class used for kernels, and the `DeviceFunction` class is more analagous to the `_Kernel` class. Eventually `Dispatcher` will subsume these classes as we move towards a more "modern" dispatcher implementation in the CUDA target.

Tests are added; these are all skipped on the simulator because overloading doesn't really exist as a concept without compilation.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
This failure: https://dev.azure.com/numba/ff1fe4d0-ed73-4f1c-b894-1d50a27e048f/_apis/build/builds/8536/logs/130 AFAICT is nothing to do with this PR, fixed in https://github.com/numba/numba/pull/6945

**Comment by gmarkall** [comment]
A quick note on this, mainly for @stuartarchibald: Earlier versions of this had `dispatcher_registry[hardware_registry["cuda"]] = Dispatcher` in `initialize_all()`, but I removed it for this patch because it wasn't used for `@overload` - it will need adding back sometime. My understanding was that this is only needed for the `@jit` decorator with a `hardware` / `target` (whatever we end up calling it) kwarg.

**Comment by stuartarchibald** [comment]
> A quick note on this, mainly for @stuartarchibald: Earlier versions of this had `dispatcher_registry[hardware_registry["cuda"]] = Dispatcher` in `initialize_all()`, but I removed it for this patch because it wasn't used for `@overload` - it will need adding back sometime. My understanding was that this is only needed for the `@jit` decorator with a `hardware` / `target` (whatever we end up calling it) kwarg.


Agree, I think the registrations needed to get `overload` working at present can avoid having a dispatcher registered.

**Comment by gmarkall** [comment]
@stuartarchibald As you suggested OOB I've replaced the printing and captured output in the tests with a value-based check - this is much tidier, and easier to deal with issues.

File: numba/cuda/tests/cudapy/test_overload.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+        expected = CUDA_CALLS_HARDWARE_OL * CUDA_HARDWARE_OL
+        self.check_overload(kernel, expected)
+
+    def test_hardware_overloaded_calls_hardware_overloaded(self):
+        def kernel(x):
+            hardware_overloaded_calls_hardware_overloaded(x)
+
+        expected = CUDA_HARDWARE_OL_CALLS_HARDWARE_OL * CUDA_HARDWARE_OL
+        self.check_overload(kernel, expected)
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_overload.py:286]: In this section, maybe test that the CPU calls would pick up the "generic" variants?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @gmarkall, great to see this working :) Once minor comment to resolve else looks good.

File: numba/cuda/tests/cudapy/test_overload.py (hunk 1 above)
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_overload.py]: Added.

**Comment by stuartarchibald** [comment]
Thanks for adding the extra tests in f1b54d0, this looks good. There's a conflict to resolve against mainline and once done this can be tested on the farm. Thanks again.

**Comment by gmarkall** [comment]
@stuartarchibald Many thanks, conflict resolved.

**Review by stuartarchibald — APPROVED** [review]
Thanks for getting this working, it's great to see the new API in action!

**Comment by stuartarchibald** [comment]
> @stuartarchibald Many thanks, conflict resolved.

Thanks, patch approved.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_49`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_49`.

Passed.

**Pull request merged**